web/src/muted_topics_ui.js

```diff
@@ -10,37 +8,24 @@ import * as sub_store from "./sub_store";
 import * as unread_ui from "./unread_ui";
 import * as user_topics from "./user_topics";
 
-export function rerender_for_muted_topic(old_muted_topics) {
+export function handle_topic_updates(user_topic_event) {
+    // Update the UI after changes in topic visibility policies.
+    user_topics.set_user_topic(user_topic_event);
+    popover_menus.get_topic_menu_popover()?.hide();
+
     stream_list.update_streams_sidebar();
+    unread_ui.update_unread_counts();
     message_lists.current.update_muting_and_rerender();
     if (message_lists.current !== message_lists.home) {
         message_lists.home.update_muting_and_rerender();
     }
     if (overlays.settings_open() && settings_muted_topics.loaded) {
         settings_muted_topics.populate_list();
     }
-
-    // We only update those topics which could have been affected, because
-    // we want to avoid doing a complete rerender of the recent topics view,
-    // because that can be expensive.
-    const current_muted_topics = user_topics.get_user_topics_for_visibility_policy(
-        user_topics.all_visibility_policies.MUTED,
-    );
-    const maybe_affected_topics = _.unionWith(old_muted_topics, current_muted_topics, _.isEqual);
-
-    for (const topic_data of maybe_affected_topics) {
-        recent_topics_ui.update_topic_is_muted(topic_data.stream_id, topic_data.topic);
-    }
-}
```

Comment:
I rewrote this part of the implementation, because it didn't make any sense to do this diffing process when the user topic event we were passed had the information of which rows need to be updated.